### ISSUE TITLE
Systemd error message gets printed over Yuigreet interface

### ISSUE THREAD

**Issue by gnull** [apognu/tuigreet#68]
The `systemd-networkd-wait-online` service often fails on my machine on startup. When it does so, Systemd writes

> systemd[1]: Failed to start Wait for Network to be Configured.

to the system journal and that gets printed to some of the consoles (not sure which ones and how it works). If this happens when I'm in Tuigreet, the messages gets printed right over the Tuigreed interface. In my case the message fills the password field (that's where the cursor is).

Could Tuigreet somehow disable printing of system log messages to the console it works in? (And maybe reenable it when Tuigreet quits.)

I'm using `tuigreet 0.7.3` running Archlinux.

**Comment by gnull** [comment]
There's a chance this is not an issue for tuigreet, since it's logger daemons that are writing stuff to console. Users probably need to configure their `systemd-journald` to stop writing messages to console. After some quick googling, I couldn't find a way to disable the messages on just one specific console.

I would be nice though to have a paragraph in tuigreet docs explaining this.

**Comment by UltraBlackLinux** [comment]
I am facing this overflow issue since I removed the quiet option from grub. I really value seeing a log, so I won't disable that. I hope it can be fixed somehow.

**Comment by Occhioverde** [comment]
Personally, I am afraid that this change in Greetd's configuration does not really solve the problem.

My `/etc/greetd/config.toml` file is something like

```toml
[terminal]
vt = "next"

[default_session]
command = "tuigreet --remember --time --issue --asterisks --cmd sway"
user = "greeter"
```

while this is what Tuigreet looks like when I boot without the kernel quiet parameter:

![SystemD Log bleeding in Tuigreet](https://user-images.githubusercontent.com/18429412/177621610-804e0672-861e-44d9-8f8b-d75b2cdc87ce.jpg)

**Comment by gnull** [comment]
You're right. I just saw some error messages with `vt = "next"` and `quiet` in kernel parameters. So the issue still exists.

**Comment by apognu** [comment]
I feel you, and I have the same issue from time to time. Unfortunately, I have no idea if and how this could be solved, the same issue happens to me with any console program I invoke.

Short of rewriting the whole screen, including whitespace, at every frame, I don't see a solution.

If anyone has an idea within the scope of `tuigreet`, I'm all ears, but for now, I'm blank. A root program, if it wants to write on a tty, will write on a tty, there is little I can do.

**Comment by UltraBlackLinux** [comment]
It might be worth looking at [ly](https://github.com/fairyglade/ly/blob/master/src/draw.c). They are kinda doing the same thing as tuigreet but different, and they don't have the problem described here.

I have no experience in using C, but you might be able to find what updates their screen. My best  guess currently is the very bottom of the file

**Comment by foopsss** [comment]
I followed a [suggestion](https://bbs.archlinux.org/viewtopic.php?id=198706) given on Arch forums and the log messages on tuigreet seem to be gone.

Basically, redirect all the logs to a different TTY than the one you are using to boot tuigreet.

**Comment by gnull** [comment]
Lucas,

Do the journald messages still go away when you remove `quiet` kernel
commandline argument? (I.e. remove it from /etc/default/grub, do
grub-mkconfig and reboot.)

For me, the above still produces the messages over tuigreet
interface. Although with `quiet` set I don't see anything (maybe I
haven't testing it enough).

Ivan

**Comment by Occhioverde** [comment]
Finally, after reading lots of threads and trying out different possible solutions, I found out that SystemD stops printing its status messages (like it was started with the quiet parameter) upon receiving `SIGRTMIN+21` and then restarts doing it when a `SIGRTMIN+20` signal is sent (Source: https://man.archlinux.org/man/systemd.1#SIGNALS).

So I created a file named `/usr/lib/systemd/system/greetd.service.d/00-nobleed.conf`, I pasted into it the following content and restarted my machine:
```ini
[Service]
ExecStartPre=kill -SIGRTMIN+21 1
ExecStopPost=kill -SIGRTMIN+20 1
```

Thanks to this modification, SystemD stops printing its log before launching greetd (and tuigreet along with it) restarting it only after the login screen is closed.

I'm not entirely sure that this solution can stop error messages too from being printed over tuigreet, but for now this is the best workaround I was able to find.

**Comment by foopsss** [comment]
> Lucas,
> 
> Do the journald messages still go away when you remove `quiet` kernel
> commandline argument? (I.e. remove it from /etc/default/grub, do
> grub-mkconfig and reboot.)
> 
> For me, the above still produces the messages over tuigreet
> interface. Although with `quiet` set I don't see anything (maybe I
> haven't testing it enough).
> 
> Ivan
> 

Hello.

It seems I rushed a bit with my first comment, because the log messages came back after a full reboot with the workaround I posted before. My apologies.

I also tried what you asked me to do. After three reboots, I've seen no log messages over the tuigreet prompt every time I tried to log in. However, I did not have the quiet flag set in my GRUB config before.

I'll wait a few more days, and, if the log messages come again, I'll try the workaround proposed by occhioverde and report back.

My tuigreet version is 0.7.2. Cheers.

**Comment by DotIN13** [comment]
I tried @Occhioverde's solution and it seemed to have worked. The logs and errors are gone now, will report again if it comes back.

**Comment by Lorilandly** [comment]
@apognu What about adding an option, for example by pressing F5, to allow user to manually update the screen? 
Screen bleeding is a common issue for tui applications, and many solves it by allowing a manual update, e.g., vim refresh the screen with CTL_L.

**Comment by literal-line** [comment]
![IMG_20221115_114307.jpg](https://user-images.githubusercontent.com/13988479/201977105-39faad11-e62d-4d7e-8fb5-d8d3c92e117d.jpg)

I'm having a similar issue. I'll try the fixes above and see if they work.

**Comment by DotIN13** [comment]
> I tried @Occhioverde's solution and it seemed to have worked. The logs and errors are gone now, will report again if it comes back.
> 
> EDIT: @foopsss's suggestion I also tried, but no luck though.

After sometime, I reverted the changes from @Occhioverde's solution as I happened needed to check the systemd logs at the login screen as my system won't boot into desktop environment. So I had to boot into a livecd and delete the lines in the service file. I guess it's just hard to get them all, debuggability and cleanliness.

**Comment by Yakkhini** [comment]
Here is a more elegant way to fix this problem. Inspired by @UltraBlackLinux and I checked ly's repo to find how did they fix this issue. In [#4ab41f0](https://github.com/fairyglade/ly/commit/4ab41f0bb9c137cc9b7c47b772f14503e87c209b) I found that they just change service file with "Type = idle".

In [getty - ArchWiki](https://wiki.archlinux.org/title/Getty#Virtual_console), 2.3.1, a small tip point out how it works:

> The option `Type=idle` found in the default `getty@.service` will delay the service startup until all jobs (state change requests to units) are completed in order to avoid polluting the login prompt with boot-up messages. When [starting X automatically](https://wiki.archlinux.org/title/Start_X_at_login), it may be useful to start `getty@tty1.service` immediately by adding `Type=simple` into the drop-in file. Both the init system and startx can be [silenced](https://wiki.archlinux.org/title/Silent_boot) to avoid the interleaving of their messages during boot-up.
>
> —— Tip, 2.3.1, getty - ArchWiki

Then I add a new line "Type=idle" in greetd.service file.

![yakkhini@20221231161109](https://user-images.githubusercontent.com/59007159/210130148-4bb35b7f-ff66-4fee-9201-64bb93cfb27c.png)

And tuigreet looks fine after restart.

**Comment by cenktaskin** [comment]
> And tuigreet looks fine after restart.

I wonder if I'm alone in this, but this doesn't solve the issue for me.

**Comment by Yakkhini** [comment]
> > And tuigreet looks fine after restart.
> 
> I wonder if I'm alone in this, but this doesn't solve the issue for me.
> 
> I saw the exact error msgs as @literal-line above.

On my machine, the same message appeared 2 to 3 times last month and did not reappear this month. I had been using ly for about a year and this problem occurred similarly, but very rarely.


This approach ensures that tuigreet has roughly the same effect as ly, which is acceptable to me personally in terms of the tradeoff between simplicity and effectiveness. However, I think everyone has different expectations for their machines, and different performance, packages, and configurations of their machines, so this may not be the optimal solution for you.


For example, if the issue you describe is recurring on every boot, then I suggest you consider the approach of others in this issue, which may be more satisfactory to you.


:)


(The above was translated by DeepL Translator)

在我的机器上，同样的信息在上个月出现了 2~3 次，这个月没有再出现。我曾使用 ly 大约一年，这个问题也同样发生过，但非常稀少。


这一方法保证了 tuigreet 与 ly 的效果大致相同，在简易性与效果的权衡来看我个人是可以接受的。但我想每个人对自己机器的期待不同，机器的性能、软件包、配置也不相同，所以这问题对你可能不是最优解。


比如，如果你所说的问题在每一次启动中均复现，那么我建议你还是考虑选用本 issue 中其他人的方法，或许会让你更满意。


:)

**Comment by paulolieuthier** [comment]
Setting `Type=idle` in greetd.service file gets rid of Systemd logs, but there is still log from the kernel. What fixed it for me was to reduce the kernel log level, as explained here: https://superuser.com/questions/351387/how-to-stop-kernel-messages-from-flooding-my-console.

**Comment by sjcobb2022** [comment]
I found this config from some thread on reddit. 

```nix
  # this is a life saver.
  # literally no documentation about this anywhere.
  # might be good to write about this...
  # https://www.reddit.com/r/NixOS/comments/u0cdpi/tuigreet_with_xmonad_how/
  systemd.services.greetd.serviceConfig = {
    Type = "idle";
    StandardInput = "tty";
    StandardOutput = "tty";
    StandardError = "journal"; # Without this errors will spam on screen
    # Without these bootlogs will spam on screen
    TTYReset = true;
    TTYVHangup = true;
    TTYVTDisallocate = true;
  };
```

**Comment by farbeiza** [comment]
As an idea, why not making <ESC> clear and redraw the whole screen?

**Comment by Chess3D** [comment]
I was having the same problem in Arch and fixed it using by adding the below systemd config with `systemctl edit greetd --drop-in=quite`

```
# /etc/systemd/system/greetd.service.d/quite.conf
[Unit]
After=getty@tty2.service

[Service]
Type=idle
StandardInput=tty
TTYPath=/dev/tty2
TTYReset=yes
TTYVHangup=yes
```

**Comment by apognu** [comment]
A [comment on Reddit](https://www.reddit.com/r/NixOS/s/JYd6Q9MZ64) seems to have had some luck with these `systemd` directives in the `greetd` service:

```
Type=idle
StandardInput=tty
StandardOutput=tty
StandardError=journal
TTYReset=true
TTYVHangup=true
TTYVTDisallocate=true
```

I can't really reproduce the issue on my systems, but if someone wants to give it a try, have at it.

**Comment by sjcobb2022** [comment]
> A [comment on Reddit](https://www.reddit.com/r/NixOS/s/JYd6Q9MZ64) seems to have had some luck with these `systemd` directives in the `greetd` service:
> 
> ```
> Type=idle
> StandardInput=tty
> StandardOutput=tty
> StandardError=journal
> TTYReset=true
> TTYVHangup=true
> TTYVTDisallocate=true
> ```
> 
> I can't really reproduce the issue on my systems, but if someone wants to give it a try, have at it.

This is what I have in my system config. It works very well and I do not have any spam on my screen.

**Comment by deviantsemicolon** [comment]
> I found this config from some thread on reddit. For working with NixOS.
> 
> Here is the original [post](https://www.reddit.com/r/NixOS/comments/u0cdpi/tuigreet_with_xmonad_how/). I've been using it in my config and it works like a charm.
> 
> ```nix
>   # this is a life saver.
>   # literally no documentation about this anywhere.
>   # might be good to write about this...
>   # https://www.reddit.com/r/NixOS/comments/u0cdpi/tuigreet_with_xmonad_how/
>   systemd.services.greetd.serviceConfig = {
>     Type = "idle";
>     StandardInput = "tty";
>     StandardOutput = "tty";
>     StandardError = "journal"; # Without this errors will spam on screen
>     # Without these bootlogs will spam on screen
>     TTYReset = true;
>     TTYVHangup = true;
>     TTYVTDisallocate = true;
>   };
> ```
> 
> Edit: An example config for arch. I actually haven't tested this one but based on other systemd configs i've seen it should work.
> 
> ```toml
> # /etc/systemd/system/greetd.d/no_spam.conf
> [Service]
> Type = "idle"
> StandardOutput = "tty"
> # Without this errors will spam on screen
> StandardError = "journal"
> # Without these bootlogs will spam on screen
> TTYReset = true
> TTYVHangup = true
> TTYVTDisallocate = true
> ```
> 
> The arch wiki says `systemctl edit unit --drop-in=drop_in_name` is the easiest way to create new configs for your systemd

the nix thing works perfectly. thank you so much

**Comment by tmccombs** [comment]
Is there any chance this could merged into greetd? 

Or maybe have tuigreet send the signals described in https://github.com/apognu/tuigreet/issues/68#issuecomment-1192683029 to systemd as part of tuigreet itself?

**Comment by apognu** [comment]
I do not think this should be the responsibility of `tuigreet` to do that, unfortunately.

**Comment by tmccombs** [comment]
Another possible solution is to set `console=tty1` on the kernel commandline so that console output only goes to tty1  instead of all ttys. 

I haven't confirmed this actually fixes the problem yet though.

**Comment by johanbluecreek** [comment]
I tried the override dropins on Gentoo with systemd, and it does not solve the problem. The particular journal entries that show up for me are the messages from the docker daemon starting containers.

**Comment by sjcobb2022** [comment]
> I tried the override dropins on Gentoo with systemd, and it does not solve the problem. The particular journal entries that show up for me are the messages from the docker daemon starting containers.

Could start greetd.service after docker. Would add a little bit of boot time but remove the logs.

**Comment by apognu** [comment]
Even though I sympathize with people encountering this issue, I do not think this is `tuigreet`'s fault and am unsure if I can prevent that. After all, the Linux consoles are shared resources, and if things are writing to it, they will appear.

I will not lock discussion but will close this.

I will still monitor responses just in case something comes up.

**Comment by Yakkhini** [comment]
Glad to see this issue finally closed and the typo in title fixed. ;)

Thanks for concerning from apognu and all you guys.

**Comment by medv** [comment]
It was consistently fixed with the above solution, but has returned after several updates.

This is a fine outcome. Knowing it is not a simple issue and not the result of my incompetence, it's really not that much of an eyesore. tuigreet is still a net positive in my config and morning routine, thank you @apognu.

**Comment by MagneticNeedle** [comment]
> Another possible solution is to set `console=tty1` on the kernel commandline so that console output only goes to tty1 instead of all ttys.
> 
> I haven't confirmed this actually fixes the problem yet though.
> 
> Edit: This doesn't seem to work. Although I don't know why.

i have this, and it works fine

```bash
# /etc/default/grub (ignore the nvidia stuff)
GRUB_CMDLINE_LINUX_DEFAULT="console=tty1 loglevel=3 modprobe.blacklist=nouveau nvidia_drm.modeset=1 nvidia-drm.fbdev=1"

# /proc/sys/kernel/printk
3	4	1	4
```

**Comment by Renkoto** [comment]
> # /etc/default/grub (ignore the nvidia stuff)
> GRUB_CMDLINE_LINUX_DEFAULT="**console=tty1 loglevel=3** modprobe.blacklist=nouveau nvidia_drm.modeset=1 nvidia-drm.fbdev=1"
> 
> # /proc/sys/kernel/printk
> 3	4	1	4

Artix user here, this fixed my issue with tuigreet + consolefont.

**Comment by rubaboo** [comment]
Just adding some info, confirming that `loglevel=3` seems to be an acceptable workaround.
1. I think [this](https://unix.stackexchange.com/a/592897) explains why `vt = next` config or `console=` kernel parameter don't help:
> the KVT subsystem by default writes all console output, when it has been selected as a console in the first place, to the currently active KVT.

2. And [this](https://wiki.archlinux.org/title/Silent_boot#Kernel_parameters) assures that
> The loglevel parameter will only change that which is printed to the console, the levels of dmesg itself will not be affected and will still be available through the journal as well as dmesg.

**Comment by jsskeen-utah** [comment]
I fully understand why this is not a `tuigreet`-caused issue, but I wondered if it would be possible to have it when the user presses ESC, it not only clears the form but also clears and redraws the screen (or some other keybind for that functionality). That way no matter if/when things print over the UI, the user can quickly restore what they want (or not, and just look at the messages that have been logged out over tuigreet, their choice).

Edit: I asked my AI coding agent what it would take to do this. Here is its response:

> **Files to change:** `crates/tuigreet/src/keyboard.rs`, `crates/tuigreet/src/event.rs`, `crates/tuigreet/src/main.rs`
>
> **Changes:**
> 1. **event.rs**: Added a new `ClearScreen` event variant
> 2. **keyboard.rs**: ESC handler now sends `Event::ClearScreen` (in addition to cancelling auth and resetting the form)
> 3. **main.rs**: `ClearScreen` handler calls `terminal.clear()` then redraws the full UI
>
> This forces a full terminal clear and redraw on ESC, which covers up any noisy text that bled through onto the screen. The form reset behavior is unchanged.

The implementation is on the `esc-clear-screen` branch of my fork if anyone wants to test it. I would be happy to open a PR for it (or feel free to implement it yourself). I will probably try this locally and see if it helps, because I have a pretty noisy TTY7 for whatever reason, and I do not necessarily want to silence everything because there is a small chance I would want to see what it is logging.

Edit 2: I just rebooted and tried it out - the ESC worked like a charm! There was a slight delay, but I think it was just fine. I'm going to open a PR for this.

**Comment by jsskeen-utah** [comment]
I've opened up #226 . It's working reliably with the unit/integration tests and `--mock` mode. Essentially it just overwrites ratatui's screen buffer cache with the `█` character so that it is forced to redraw everything. This only happens on ESC press so it's not going to cause performance issues when idle.

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/7cc97795-c0ff-4871-915e-8a0ca6f0c3b1" />
